### ISSUE TITLE
feat: selectors for element persistence

### ISSUE THREAD

**Issue by gdi2290** [angular/preboot#21]
allow selectors for elements that need to be transferred over from preboot to the angular rendered view.

use case: ADs

**Comment by jeffwhelpley** [comment]
Not really sure what this means. @gdi2290 can you provide a little more context.

**Comment by jeffwhelpley** [comment]
oh, you mean hydration?

**Comment by FrozenPandaz** [comment]
+1

**Comment by gdi2290** [comment]
the server rendered view needs to detach elements into a registry by element id then reattach the element via component during angular bootstrap. This can also be done outside of preboot manually by the developer.
```typescript
@Directive({
   selector: '[universalAd],[universal-ad]'
})
class UniversalAd {
  constructor(@Attribute('id') public id: string, public adRegistry: MyAdRegistry, public elementRef: ElementRef, cdRef: ChangeDetectorRef) {
      if (this.adRegistry.isRegistered(id)) {
        this.adRegistry.reattach(id, this.elementRef.nativeElement);
      } else {
        this.adRegistry.registerElement(id, this.elementRef.nativeElement);
      }
        cdRef.detach();
  }
  ngOnDestory() {
    this.adRegistry.detach(this.id, this.elementRef.nativeElement);
  }
}


@Component({
  template: `
  <div universalAd id="importantAd"></div>
`
})
class App {
  constructor() {

  }
}

@NgModule({
  bootstrap: [App],
  declarations: [ App, UniversalAd ],
  imports: [
    UniversalModule
  ],
  providers: [
    {
      provide: MyAdRegistry,
      useFactory: () => {
        var listOfAdIds = {
          'importantAd': document.getElementById('importantAd')
        }
        var listOfClientElements = {}
        return {
          isRegistered: (id) => Boolean(listOfClientElements[id]),
          registerElement: (id, el) => listOfClientElements[id] = el,
          complete() { 
            Object.keys(listOfClientElements).forEach(id => {
              // todo: use raf
              var ngEl = listOfAdIds[id];
              var serverEl = listOfClientElements[id];
              document.removeChild(serverEl); 
              ngEl.appendChild(serverEl)
            });
          },
          detach(id, el) {
            var serverEl = listOfClientElements[id];
            el.removeChild(serverEl)
          },
          reattach(id, el) {
            var serverEl = listOfClientElements[id];
            el.appendChild(serverEl)
          }
        };
      }
    }
  ]
})
class MainModule {
}


platformUniversalDynamic()
.bootstrap(MainModule)
.then(moduleRef => {
  const adRegistry = moduleRef.injector.get(MyAdRegistry)
  adRegistry.complete();
  preboot.complete();
  return moduleRef;
})
```

**Comment by FrozenPandaz** [comment]
If I read that correctly, it looks like you're trying to just appendChild (The Javascript way of moving elements) with the element from a server view/ registry onto the client view?

That breaks googletag ads and I think most iframes.

**Comment by jeffwhelpley** [comment]
I am going to close this because transferring elements is outside the scope of preboot for now. I think the ultimate solution for this is a hydration solution which we can work out outside of preboot.
